File: Striver/JAVA/longest_consecutive_sequence_in_array.java
// Given an array nums of n integers.

// Return the length of the longest sequence of consecutive integers.
// The integers in this sequence can appear in any order.

// Look For Reference In the Book-2 On Pg No 132-133


import java.util.HashSet;

public class longest_consecutive_sequence_in_array {
    public static int longest_sequence(int[] nums){
        if(nums==null||nums.length==0)  return 0;
        int longest = 1;
        HashSet<Integer> set = new HashSet<>();
        for(int num : nums){
            set.add(num);
        }
        for(int num : nums){
            if(!set.contains(num-1)){
                int cnt = 1;
                int curr = num;
                while(set.contains(curr+1)){
                    cnt++;
                    curr++;
                }
                longest = Math.max(cnt, longest);
            }
        }
        return longest;
    }
    public static void main(String[] args) {
        int[] arr = new int[]{0, 3, 7, 2, 5, 8, 4, 6, 0, 1};
        System.out.println("The Longest Consecutive Sequence In Array Is : "+longest_sequence(arr));
    }
}

File: Striver/CPP/longest_consecutive_sequence_in_array.cpp
// Given an array nums of n integers.

// Return the length of the longest sequence of consecutive integers.
// The integers in this sequence can appear in any order.

// Look For Reference In the Book-2 On Pg No 132-133


#include<iostream>
#include<vector>
#include<unordered_set>
#include<algorithm>

using namespace std;

int longest_sequence(vector<int>& nums){
    // if(nums==nullptr||nums.size()==0)   return 0;
    if(nums.size()==0)   return 0;
    int longest = 1;
    unordered_set<int> set;
    for(int num : nums){
        set.insert(num);
    }
    for(int num : nums){
        if(set.find(num-1)==set.end()){
            int cnt = 1;
            int curr = num;
            while(set.find(curr+1)!=set.end()){
                cnt++;
                curr++;
            }
            longest = max(cnt, longest);
        }
    }
    return longest;
};

int main(){
    vector<int> nums = {100, 4, 200, 1, 3, 2};
    cout<<"The Longest Consecutive Sequence Is : "<<longest_sequence(nums)<<endl;
    return 0;
}